asmcomp/cmmgen.ml

```diff
@@ -1660,11 +1660,50 @@ let make_switch arg cases actions dbg =
       | Cconst_symbol (s, _) -> Csymbol_address s
       | _ -> assert false in
     let const_actions = Array.map to_data_item actions in
-    let table = Compilenv.new_const_symbol () in
-    Cmmgen_state.add_constant table (Const_table (Local,
-        Array.to_list (Array.map (fun act ->
-          const_actions.(act)) cases)));
-    addr_array_ref (Cconst_symbol (table, dbg)) (tag_int arg dbg) dbg
+    let linear_function_result =
+      (* In case the resulting integers are a linear function of the index, we
+         don't emit a table, and just compute the result directly *)
+      let length = Array.length cases in
+      if length >= 2
+      then begin
+        match const_actions.(cases.(0)), const_actions.(cases.(1)) with
+        | Cint v0, Cint v1 ->
```

Comment:
The use of the reference just a Boolean is not so nice.

````ocaml
let slope = Nativeint.sub v1 v0 in
let check i idx = function
  | Cint v -> v = Nativeint.(add v0 (mul (of_int i) slope))
  | _ -> false
in
if Misc.Stdlib.Array.forall_i (fun i idx -> check i const_actions.(idx)) cases then Some (v0, slope)
else None
````
(and add the definition of `Misc.Stdlib.Array.forall_i `)

Reply:
Done, and I added the definition to Stdlib, not Misc.Stdlib, because it seems generally useful.
I have no issue moving it to Misc if that's not a good idea.

Reply:
I would recommend moving to `Misc` and then opening a separate PR about the move to the stdlib. Any stdlib addition is going to need a lot of scrutiny and public discussion, you don't want to have this as a dependency of your PR.